### PR TITLE
Implement saving

### PR DESCRIPTION
**Pull request by Aegyo**
Majority of changes here are updating other files to reference the instance of Player (named player) instead of the class.

The only property not moved into the instance is Player.routeKillsNeeded, I'm not sure why this is observable, but as far as I know it will never change. Unless there are plans for this I don't know about yet, it might be best to move this into GameConstants.

I haven't encountered any issues with saving or loading in my limited testing. If you want to test it out, there is a Save.reset() function which will clear out the localStorage and reset your browser, just in case something breaks (or you want to test loading the game when there is no saved player)

### CODE REVIEW

**Comment by Ishadijcks** [comment]
`Player.routeKillsNeeded` can be decreased to 7 as a prestige perk

**Comment by Aegyo** [comment]
Ah ok, I'll move that into the constructor also then.

File: src/scripts/Save.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    public static filter(object: any, keep: string[]): Object {
+        let filtered = {}, prop;
+        for (prop in object) {
+            if (keep.indexOf(prop) > -1) {
+                filtered[prop] = object[prop]
+            }
+        }
+        return filtered
+    }
+
```
Comment (Ishadijcks) [inline src/scripts/Save.ts]: Clever, I like it

Comment (Aegyo) [inline src/scripts/Save.ts]: Should I add a comment explaining the parameters of that function, or is it fairly clear what they should be?

Comment (Ishadijcks) [inline src/scripts/Save.ts]: Yeah, would be nice!

**Comment by Ishadijcks** [comment]
Getting a typescript compile error:
```src\scripts\Player.ts(36,13): error TS2322: Type 'KnockoutObservableArray<{}>' is not assignable to type 'KnockoutObservableArray<string>'.```

at 
```this._caughtShinyList = ko.observableArray(savedPlayer._caughtShinyList);```

**Comment by Aegyo** [comment]
Strange, I don't get that error. Any idea what is causing that?

**Comment by Ishadijcks** [comment]
Also... 
image

**Comment by Aegyo** [comment]
Yeah I just noticed that too, it's new from that last commit. I'll close this until I fix it and make sure there are no other bugs.

**Comment by Ishadijcks** [comment]
Calling `Save.reset()`
Routekills works,
Let the game save, 
Refresh browser,
NaN Routekills

**Comment by Aegyo** [comment]
I'm still not getting that compile error, even when restarting gulp.

**Comment by Ishadijcks** [comment]
Saving seems to work now, but before I merge I'm going to search for the cause of the error.

**Comment by Ishadijcks** [comment]
`ko.observableArray(savedPlayer._caughtShinyList);`
to
`ko.observableArray<string>(savedPlayer._caughtShinyList);`
fixes the problem.

I guess my WebStorm is somehow stricter than yours or something?

**Pull request merged**